mailpoet/lib/Automation/Engine/Exceptions.php

```diff
@@ -127,16 +133,66 @@ public static function subjectLoadFailed(string $key, array $args): InvalidState
       ->withMessage(sprintf(__('Subject with key "%1$s" and args "%2$s" failed to load.', 'mailpoet'), $key, Json::encode($args)));
   }
 
-  public static function multipleSubjectsFound(string $key): InvalidStateException {
+  public static function subjectDataNotFound(string $key, int $workflowRunId): NotFoundException {
+    return NotFoundException::create()
+      ->withErrorCode(self::SUBJECT_DATA_NOT_FOUND)
+      // translators: %1$s is the key of the subject, %2$d is workflow run ID.
+      ->withMessage(
+        sprintf(__("Subject data for subject with key '%1\$s' not found for workflow run with ID '%2\$d'.", 'mailpoet'), $key, $workflowRunId)
+      );
+  }
+
+  public static function multipleSubjectsFound(string $key, int $workflowRunId): InvalidStateException {
     return InvalidStateException::create()
       ->withErrorCode(self::MULTIPLE_SUBJECTS_FOUND)
-      // translators: %s is the name of the key.
-      ->withMessage(sprintf(__("Multiple subjects with key '%s' found, only one expected.", 'mailpoet'), $key));
+      // translators: %1$s is the key of the subject, %2$d is workflow run ID.
+      ->withMessage(
+        sprintf(__("Multiple subjects with key '%1\$s' found for workflow run with ID '%2\$d', only one expected.", 'mailpoet'), $key, $workflowRunId)
+      );
+  }
+
+  public static function payloadNotFound(string $class, int $workflowRunId): NotFoundException {
+    return NotFoundException::create()
+      ->withErrorCode(self::PAYLOAD_NOT_FOUND)
+      // translators: %1$s is the class of the payload, %2$d is workflow run ID.
+      ->withMessage(
+        sprintf(__("Payload of class '%1\$s' not found for workflow run with ID '%2\$d'.", 'mailpoet'), $class, $workflowRunId)
+      );
+  }
+
+  public static function multiplePayloadsFound(string $class, int $workflowRunId): NotFoundException {
+    return NotFoundException::create()
+      ->withErrorCode(self::MULTIPLE_PAYLOADS_FOUND)
+      // translators: %1$s is the class of the payloads, %2$d is workflow run ID.
+      ->withMessage(
+        sprintf(__("Multiple payloads of class '%1\$s' found for workflow run with ID '%2\$d'.", 'mailpoet'), $class, $workflowRunId)
+      );
   }
 
   public static function workflowStructureModificationNotSupported(): UnexpectedValueException {
     return UnexpectedValueException::create()
       ->withErrorCode(self::WORKFLOW_STRUCTURE_MODIFICATION_NOT_SUPPORTED)
       ->withMessage(__("Workflow structure modification not supported.", 'mailpoet'));
   }
+
+  public static function workflowStructureNotValid(string $detail): UnexpectedValueException {
+    return UnexpectedValueException::create()
+      ->withErrorCode(self::WORKFLOW_STRUCTURE_NOT_VALID)
+      // translators: %s is a detailed information
+      ->withMessage(sprintf(__("Invalid workflow structure: %s", 'mailpoet'), $detail));
+  }
```

Comment:
It can be pretty hard to determine, which rule was actually broken. Once we debug on a translated site, it would be even harder. Maybe we could have an error code per rule?

Reply:
One thing to note here is that these errors should never occur to us. Our frontend should never construct a workflow that would break those rules. If that happens, we have a bug. So it's mainly to protect against bugs and people playing with the API directly.

That said, we could add an error per rule... or maybe just some specific `rule` field in the `data`?

Reply:
> or maybe just some specific rule field in the data

Whatever suits you best there. I do not have a strong opinion as long as we can quickly identify the broken rule.

Reply:
I think I prefer just adding the rule ID, but I'd move it to the follow-up ticket, if possible. Currently, we don't have an easy way of extending the error response data when converted from an exception but I'll probably need that in the follow-up ticket anyway.

Reply:
sounds good to me.